MTSRE-1447: Handled manual install plan approval

pepedocs opened the pull request:
### What type of PR is this?
_feature_

### What this PR does / why we need it?
It sets the olm reconciler to report a more accurate status when a subscription's install plan is pending due to it waiting for a manual approval.

Update
1. added getters to enhance readability

### Which Jira/Github issue(s) this PR fixes?

_Fixes #[MTSRE-1447](https://issues.redhat.com//browse/MTSRE-1447)_

### Special notes for your reviewer:
- This is my first ADO commit
- I'm an exp. programmer but new in Go
- Please guide as necessary

### Pre-checks (if applicable):
- [ ] Tested latest changes against a cluster
- [x] Ran `make test-unit` command locally to run all the unit tests and mock tests locally.
- [ ] Included documentation changes with PR
- [x] Squashed your commits

pepedocs commented:
Hi @Ajpantuso or @ashishmax31 PTAL when you get a chance.

gvanderpotte commented on internal/controllers/addon/phase_ensure_subscription.go:
```diff
@@ -96,6 +96,8 @@ func (r *olmReconciler) ensureSubscription(
 		return resultRetry, client.ObjectKey{}, nil
 	}
 
+	r.observedSubscription = *observedSubscription
```
nit: out of an habit, it stresses me out when I see a pointer dereferenced without it being checked if it is nil before. checking above, where the object is created (in reconcilesubscription), it cannot be nil, unless error case, but still, maybe a comment here would help (for the sake of clarity), or nil case could be checked earlier, to be belt and braces.

pepedocs replied:
@gvanderpotte Thanks, that's just right, I forgot to check for nil.

erdii replied:
line 71: `observedSubscription, err := r.reconcileSubscription(ctx, desiredSubscription)`
observedSuscription is not nil if reconcileSubscription either returns a non-nil subscription pointer or a non-nil error.

erdii replied:
And that's the case: https://github.com/pepedocs/addon-operator/blob/fe70c681ed8bbc9da4ede8d8234600f93dbd5f33/internal/controllers/addon/phase_ensure_subscription.go#L107

pepedocs replied:
I added a nil check. For 2 reasons;

1. It's simpler to not assume about `reconcileSubscription`'s implementation.
2. There's still a possibility that observedSubscription will be nil (e.g. future code change) although unlikely. If this happens a nil check is required.

gvanderpotte reviewed (commented):
One nit, but otherwise, LGTM

venkateshsredhat commented:
Changes look good @pepedocs  . Just another approach to this :

Instead of adding observedSubscription  to the olmReconciler struct . You can try to return the observedSubscription from the ensureSubscription func in https://github.com/openshift/addon-operator/blob/820afed1588123df8ff66ef7f2f0aff78f5bb2e8/internal/controllers/addon/phase_ensure_subscription.go#L23 . 
And then you can pass the  observedSubscription into the observeOperatorResource func as an argument here https://github.com/openshift/addon-operator/blob/820afed1588123df8ff66ef7f2f0aff78f5bb2e8/internal/controllers/addon/phase_observe_operatorresource.go#L18 . 
 
I would wait for comments from Ashish and Andrew but yeah the logic looks good .

pepedocs commented:
> Instead of adding observedSubscription to the olmReconciler struct . You can try to return the observedSubscription from the ensureSubscription func in

Thanks @venkateshsredhat The reason why I didn't choose this option is because it unnecessarily adds inconsistencies between the return values and params of the phases - This does not encourage intuitive patterns in the code. Most importantly, since we are using a struct here, it essentially is for keeping state, why not use it to keep track of the objects that it has processed so far so it can be reused in other areas itself?

pepedocs commented:
@venkateshsredhat On second thought, yes I am waiting for ashish or andrew's review, I realized that I'm not really sure if olmReconciler is thread-safe so if it is not, it'd be better to choose the other option which is to return the sub instead.

Ajpantuso commented on internal/controllers/addon/olm_reconciler.go:
```diff
@@ -21,6 +21,7 @@ type olmReconciler struct {
 	client                  client.Client
 	uncachedClient          client.Client
 	operatorResourceHandler operatorResourceHandler
+	observedSubscription    operatorsv1alpha1.Subscription
```
To answer your questions @pepedocs 
1. No, `olmReconciler` is not concurrency-safe
2. There is not a 1:1 relationship between `olmReconciler` instances and addons, but rather one instance of `olmReconciler` is globally used for all addons. So what you really would want here is a concurrency-safe map from addons to subscriptions.
3. While an additional map would work the clients, caches, and informers supporting them will already prevent unnecessary additional API calls by returning cached objects and keeping them up-to-date. So an additional cache just for the `olmReconciler` would be redundant. A simple `get` call in `phaseObserveOperatorResource` would do.

pepedocs commented:
hey @venkateshsredhat @erdii @Ajpantuso @gvanderpotte can you please PTAL? I've modified the previous commits to the following:

1. Removed the usage of struct variable as it wasn't thread-safe
2. Instead use client Get to get subscription and installplan 
 
  - I've also created some Getters to enhance readability a bit and hopefully encourage reusing/making them.
 
3. I learned that there is an assumption about objects retrieved by the client and that it has some level of nil pointer safety guarantees, I do not know how to confirm this. Given my default strategy, I have checked for nils before using them. Please let me know these checks are not necessary.

venkateshsredhat commented on apis/addons/v1alpha1/addons_types.go:
```diff
@@ -486,6 +491,21 @@ func (a *Addon) UpgradeCompleteForCurrentVersion() bool {
+		if a.Spec.Install.OLMAllNamespaces != nil {
+			return a.Spec.Install.OLMAllNamespaces.AddonInstallOLMCommon, nil
+		}
+	default:
```
Just a small thing we can keep it as " case addonsv1alpha1.OLMOwnNamespace: "  . Even though I see only two cases .

pepedocs replied:
@venkateshsredhat Did you mean no default and just return an error if the install type is not supported?

venkateshsredhat replied:
@pepedocs  I see currently we handle only two cases here https://github.com/openshift/addon-operator/blob/main/internal/controllers/addon/phase_ensure_subscription.go#L38 . So I would want this function to be similar to that .

pepedocs replied:
@venkateshsredhat gotcha! In that code you refer though, it didn't handle no matched case and looks like it assumed it will always match. I'm going to update my commit but I'll also add the error part if there is not case match. Please confirm.

venkateshsredhat replied:
@pepedocs  yup , add this case and another default statement to handle error .

pepedocs replied:
@venkateshsredhat Thanks for confirming. I've removed the default case and used a specific case instead. A no-match will cause to return an error.

venkateshsredhat commented:
Looks good to me @pepedocs

Ajpantuso commented on internal/controllers/addon/phase_observe_operatorresource.go:
```diff
@@ -20,6 +20,45 @@ func (r *olmReconciler) observeOperatorResource(
 	csvKey client.ObjectKey,
 ) (requeueResult, error) {
+	if addon == nil {
+		return resultRetry, fmt.Errorf("current addon is nil")
```
Golang doesn't have runtime assertions, but this is more of an assertion than a reason to retry. If a `nil` addon was supplied to this method it would be a programmer error and should be fatal.

pepedocs replied:
Generally speaking, one nuance of passing pointers is that a nil value can be a valid value to indicate an empty var, and this depends on the assumed intention and programmer preference. I take this review as a nil addon should not even happen, therefore it must fail.

Ajpantuso replied:
Yeah essentially I think it's better to fail loudly here rather than something subtle like a retry because a retry will not resolve the deeper programming error that occurred. Again languages with assertions make it easier to add these checks in a dev build and remove them in prod builds. Since Go doesn't have that I would just recommend something more serious here like a `panic`.

Ajpantuso commented on internal/controllers/addon/phase_observe_operatorresource.go:
```diff
@@ -20,6 +20,45 @@ func (r *olmReconciler) observeOperatorResource(
+		if currentIp.Status.Phase == operatorsv1alpha1.InstallPlanPhaseRequiresApproval {
+			reportInstallPlanPending(addon)
+			// CSV will not be available at this stage therefore retry
+			return resultRetry, nil
```
We should not be retrying here since an install plan will not become approved via eventual consistency. We should return early, but we won't need to reconcile again until the `InstallPlan` is approved. Since we aren't watching `InstallPlans` that would probably be triggered by the creation of the new CSV.

Ajpantuso commented on apis/addons/v1alpha1/addons_types.go:
```diff
@@ -486,6 +491,21 @@ func (a *Addon) UpgradeCompleteForCurrentVersion() bool {
 }
 
+// Gets the AddonInstallOLMCommon by Install Type
+func (a *Addon) GetAddonInstallOLMCommon() (AddonInstallOLMCommon, error) {
```
While it is a noisy refactor this should be also used everywhere to prevent creating duplication/drift.

pepedocs replied:
Will update.

Ajpantuso reviewed (changes requested):
Looking better. Some comments to start with.

pepedocs commented on internal/controllers/addon/olm_reconciler.go:
```diff
@@ -93,3 +94,54 @@ func (r *olmReconciler) Reconcile(ctx context.Context,
+	namespace string,
+	destSub *operatorsv1alpha1.Subscription,
+) error {
+	if destSub == nil {
+		return errors.New("destination subscription is nil")
+	}
```
I prefer this. I just got the impression from the code that we prefer passing pointers (I don't prefer) and also I'm not sure about returning pointers of objects created in the stack/func in go, I just googled, go creates these objects in the heap so, I will update.

Ajpantuso replied:
Correct, the terminology in Go would be that the  object "escapes" to the heap when you return a pointer to memory you've allocated within a func since the object cannot be garbage collected as it has a lifetime beyond the scope of the func. You generally would want to avoid this and instead let the caller control allocation so they can reuse objects/buffers for optimization.

That said for little helper methods like these it's not a real concern. Nothing should escape the scope of the reconcile loop that didn't have a longer lifetime already so we aren't introducing leaks.

pepedocs commented:
Hi @Ajpantuso I've updated my commits based on your recommendations. PTAL when you get a chance. thanks.

Ajpantuso reviewed (approved):
/lgtm
